### PR TITLE
Phi-4 model bringup

### LINKED ISSUE #19418: [Bounty $500-2500] Bring up microsoft/phi-4 on Wormhole (N300/LB/QB)
# Bounty: Enable & Validate `microsoft/phi-4` on Tenstorrent Hardware

## Description
This bounty involves enabling and validating `microsoft/phi-4` on Tenstorrent’s hardware. The goal is to ensure that `microsoft/phi-4` compiles, runs end-to-end (inference), meets minimal performance benchmarks, and includes sufficient documentation for community adoption.

## Target Model & Difficulty Ratings
- **Model**:  `microsoft/phi-4`
- **Model Hardware Target**: Wormhole (N300/QB/LB)  
- **Theoretical Maximum Throughput**: 33 t/s/u (tokens per second per user)
## Difficulty Ratings (Performance Targets):
- **Easy ($500)**: Achieve ≥ 25% of theoretical maximum throughput (~8 t/s/u).
- **Medium ($1500)**: Achieve ≥ 50% of theoretical maximum throughput (~16 t/s/u).
- **Hard ($2500)**: Achieve ≥ 70% of theoretical maximum throughput (~23 t/s/u).

### 1. Functional Bring-Up
- `microsoft/phi-4` compiles and runs on N300/QB/LB with no blocking errors.

### 2. Performance Validation
- Clearly document throughput and latency.
- Achieve at least the baseline targets for the selected difficulty rating

### 3. Accuracy Validation

- Validate accuracy by running microsoft/phi-4 both on Tenstorrent hardware and a CPU baseline.

- Document comparative accuracy clearly, meeting or exceeding:
  - Achieve top-1 accuracy greater than 80% and top-5 accuracy greater than 95% when compared to CPU baseline.
  - Reference current model benchmarks here: [Tenstorrent Transformer Performance](https://github.com/tenstorrent/tt-metal/blob/main/models/tt_transformers/PERF.md)

### 3. Documentation
- Clear instructions for building, installing dependencies, and running on Tenstorrent.
- Any relevant parameter tuning or known issues noted.

---
## Bring-up Checklist 

- [ ] Model functionality
- [ ] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes
- [ ] Demo script to show the inference output
---

### Notes
- Work may involve multiple pull requests across relevant repositories (e.g., `TT-metal`, or others).
- If additional patches or dependencies are required, please include a short guide for maintainers.

### Helpful Links
- We have a [model bring up](https://github.com/tenstorrent/tt-metal/blob/ea36d1db2ea9ace0ef02cfaaaf9fce94cd91998d/models/model_bring_up.md#model-bring-up) guide that you may use for reference

### PR DESCRIPTION
**Pull request by velonica0**
### Ticket
Closes #19418

### Problem description
microsoft/phi-4 model bringup on Tenstorrent Hardware (N300)

### What's changed
In Phi-4, the WQKV matrices are mixed, and its attention mechanism uses GQA. However, `tt_transformers `simply averages the WQKV matrices.
So I distribute it according to the proportion of the head.

I improved it based on #22716 , thank you very much

### Performance Metrics

| Device      | Top-1 (%)  | Top-5 (%) |  Speed (t/s/u)  |  TTFT (ms)  |
| ----------- | ----------- |----------- | ----------- |----------- | 
| N300 | 89.50 | 99 | 28.85 | 452.48 |

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -56,62 +56,66 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
-    def prefill_forward_text(
-        self, tokens: torch.Tensor, page_table=None, kv_cache=None, prompt_lens=None, empty_slots=None
-    ):
-        if page_table is not None:
-            assert isinstance(page_table, torch.Tensor), "page_table mush be torch.Tensor"
-
+    def prefill_forward_text(self, tokens: torch.Tensor, page_table=None, kv_cache=None, prompt_lens=None):
         batch_size, batch_seq_len = tokens.shape
-        max_batch_size_per_model = self.model_args[0].max_batch_size
+        max_batch_size_per_model, batch_size_per_model = self._get_batch_size_per_model(batch_size)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: It seems like this PR is unintentionally undoing changes that are on main, seems like a rebasing issue, please check.

Comment (velonica0) [reply models/tt_transformers/tt/generator.py]: Hi, @skhorasganiTT 
The reason is that my PR depends on a PR that hasn't been merged yet #22716 . So when I make changes based on it, the tt_transformers file will revert to the version of the old PR. Should I wait for the old PR to be merged before submitting my PR, because I can't be sure if there are any other files in tt_transformers that have been modified?

Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py]: Yeah you could either wait for that PR to be merged, or you can change the base/target branch of this PR to the one from the other PR. Otherwise this one cannot be properly reviewed.

**Comment by velonica0** [comment]
@skhorasganiTT 
Could you please review the current code state? I've applied an old PR as a patch to my code, rather than making modifications on top of its branch.

File: models/tt_transformers/demo/simple_text_demo.py [hunk 2]
```diff
@@ -292,7 +292,7 @@ def prepare_generator_args(
             1,  # batch_size
             200,  # max_generated_tokens
             True,  # paged_attention
-            {"page_block_size": 64, "page_max_num_blocks_per_dp": 2048},  # page_params
+            {"page_block_size": 32, "page_max_num_blocks_per_dp": 1024},  # page_params
```
Comment (skhorasganiTT) [inline models/tt_transformers/demo/simple_text_demo.py]: These test parameters (and the ones above) should not be modified as they're shared with other models.

**Comment by skhorasganiTT** [comment]
> @skhorasganiTT Could you please review the current code state? I've applied an old PR as a patch to my code, rather than making modifications on top of its branch.

@gwangTT @mtairum can you take a look?

File: models/tt_transformers/demo/simple_text_demo.py (hunk 2 above)
Comment (velonica0) [reply models/tt_transformers/demo/simple_text_demo.py]: Done

File: models/tt_transformers/demo/simple_text_demo.py [hunk 3]
```diff
@@ -590,6 +590,18 @@ def test_demo_text(
...
+        max_context_per_device = {
+            1: 32 * 1024,
+            2: 64 * 1024,
+        }
+        max_context_supported = max_context_per_device.get(num_devices, 128 * 1024)
+        if max_context_supported < max_seq_len:
+            pytest.skip(
+                f"Max sequence length: {max_seq_len} not supported for model: {model_name_env} on device: {mesh_device}"
+            )
+
```
Comment (gwangTT) [inline models/tt_transformers/demo/simple_text_demo.py]: This piece of code should be integrated into `model_config.py`: https://github.com/tenstorrent/tt-metal/blob/190d6ac5a0594be3e0918f1edffe478f22d1f0bf/models/tt_transformers/tt/model_config.py#L1422-L1427

File: models/tt_transformers/tt/load_checkpoints.py [hunk 4]
```diff
@@ -225,15 +225,29 @@ def load_sharded_checkpoints(checkpoints, n_layers):
...
+                q_size = n_heads * head_dim
+                kv_size = n_kv_heads * head_dim
+                
+                q_tensor = tensor[:q_size]
+                k_tensor = tensor[q_size:q_size + kv_size]
+                v_tensor = tensor[q_size + kv_size:q_size + 2 * kv_size]
+            else:
+                # Default case: equal split for Q, K, V
+                q_tensor, k_tensor, v_tensor = torch.split(tensor, tensor.shape[0] // 3, dim=0)
+
```
Comment (gwangTT) [inline models/tt_transformers/tt/load_checkpoints.py]: This function, as the name suggests, is meant for splitting the keys. Currently, the tensor splitting can be moved inside the attention module -- `tt_transformers/tt/attention.py`.

File: models/tt_transformers/tt/load_checkpoints.py [hunk 5]
```diff
@@ -270,8 +284,69 @@ def convert_hf_qkv_to_meta_format(loaded_weights, head_dim):
     return converted_weights
 
 
-def convert_meta_to_hf(state_dict, head_dim):
+def fuse_mlp_meta(state_dict):
```
Comment (gwangTT) [inline models/tt_transformers/tt/load_checkpoints.py]: Similar to the comment above: moving this function to the MLP module (`tt_transformers/tt/mlp.py`) would better follow the existing code pattern.

File: models/tt_transformers/tt/load_checkpoints.py [hunk 6]
```diff
@@ -270,8 +284,69 @@ def convert_hf_qkv_to_meta_format(loaded_weights, head_dim):
...
+
+        fused_gate_up_proj = torch.vstack((wgate, wproj))
+        state_dict[f"{prefix_gate}{key_map['w_gate_up_proj']}"] = fused_gate_up_proj
+
+        del state_dict[wgate_key], state_dict[wproj_key]
+
+    return state_dict
+
+
+def fuse_qkv_meta(state_dict):
```
Comment (gwangTT) [inline models/tt_transformers/tt/load_checkpoints.py]: Similar to the comment above: moving this function to the attention module (tt_transformers/tt/attention.py) would better follow the existing code pattern.

**Review by gwangTT — COMMENTED** [review]
Flushing some comments.

File: models/tt_transformers/tt/load_checkpoints.py (hunk 4 above)
Comment (velonica0) [reply models/tt_transformers/tt/load_checkpoints.py]: image
My understanding is that the splitting of QKV weights must be done within the `split_hf_keys  `function. 

This is because in the subsequent `convert_hf_qkv_to_meta_format` function, the Q/K/V matrices are individually dimensionally transformed for [RoPE](https://github.com/tenstorrent/tt-metal/issues/25619#issuecomment-3113136274) 

Due to the characteristics of GQA , the Q, K, and V matrices must be separated before this dimensional transformation can be performed.

File: models/tt_transformers/tt/load_checkpoints.py (hunk 5 above)
Comment (velonica0) [reply models/tt_transformers/tt/load_checkpoints.py]: The `fuse_mlp_meta `and `fuse_qkv_meta `functions were designed for the `Hugging Face phi `module. Their purpose is to handle the merging of `state_dict `variables for the reference implementation, not for `tt_transformers`.

The reason for this is that in Hugging Face's Phi-3 (Phi-4) models, the QKV and MLP tensors are fused together.

In contrast, the design approach of tt_transformers is to separate the fused QKV and MLP within the split_hf_keys function.

image
image

File: models/tt_transformers/demo/simple_text_demo.py (hunk 3 above)
Comment (velonica0) [reply models/tt_transformers/demo/simple_text_demo.py]: I have removed that check.

**Comment by velonica0** [comment]
Hi, @gwangTT 
May I ask if this PR can be merged?

File: models/tt_transformers/tt/load_checkpoints.py (hunk 4 above)
Comment (gwangTT) [reply models/tt_transformers/tt/load_checkpoints.py]: Right. I am concerned that this condition:
```
if n_heads is not None and n_kv_heads is not None and n_heads != n_kv_heads
```
is not going to stop other models than GQA from going into the code block. 

Maybe it would be better to refactor the GQA specific code into a function, which is called inside `convert_hf_to_meta` after a if statement that checks the state_dict for signals for GQA? 

Or maybe the caller of `convert_hf_to_meta` is in a better position to pass in a bool value to tell `convert_hf_to_meta` whether it should run the code for GQA or not.

Comment (velonica0) [reply models/tt_transformers/tt/load_checkpoints.py]: Would it be possible for a non-GQA model to have different q_head and kv_head configurations and be mistakenly identified as a GQA model?

I have checked the config.json files for both the Qwen and Phi models, and there is no signal to indicate whether the models use GQA. 
The only method at the moment is to check the q_head and kv_head configurations.

File: models/tt_transformers/tests/reference_outputs/Phi-3-mini-128k-instruct.refpt
Comment (gwangTT) [inline models/tt_transformers/tests/reference_outputs/Phi-3-mini-128k-instruct.refpt:1]: This file should be removed from this PR.

**Comment by gwangTT** [comment]
> Hi, @gwangTT May I ask if this PR can be merged?

Sorry for the delay in responding. 

There is additional steps before we can merge this in:
- make a mirror branch that we can kick off CI runs from
- get successful CI runs to verify the accuracy and performance (or plan-b: I can run local tests to verify)
- pull from `main` and try to merge!

I can help you through the above steps. Thanks for contributing!

**Comment by velonica0** [comment]
Hi, @gwangTT 
Since the version has been updated, I opened a new #28099.
This is a mirror branch and what do I need to do next?
I have applied for permission in #28100.  Can I create a mirror branch in the main repository?

**Comment by gwangTT** [comment]
> Hi, @gwangTT Since the version has been updated, I opened a new #28099. This is a mirror branch and what do I need to do next? I have applied for permission in #28100. Can I create a mirror branch in the main repository?

Looks like I will need to create the mirror branch. I will do that now.


Actually, I left a comment on your phi-1.5 PR: https://github.com/tenstorrent/tt-metal/pull/26505#issuecomment-3255434770. Are you OK with me helping with Phi-1.4 first?

**Comment by velonica0** [comment]
@gwangTT 
I really appreciate your help with phi-4.
Can you help me create a mirror branch for #28099?

**Comment by gwangTT** [comment]
> @gwangTT I really appreciate your help with phi-4. Can you help me create a mirror branch for #28099?

#28099 only has 1 commit. Maybe it is better to reopen this one?

**Comment by velonica0** [comment]
> > @gwangTT I really appreciate your help with phi-4. Can you help me create a mirror branch for #28099?
> 
> #28099 only has 1 commit. Maybe it is better to reopen this one?

The reason is that many commits of the old PR have been merged into main(https://github.com/tenstorrent/tt-metal/pull/26210#discussion_r2265553061). 
So I only kept the commits that can be used for phi-4 in the new PR.

**Comment by gwangTT** [comment]
> > > @gwangTT I really appreciate your help with phi-4. Can you help me create a mirror branch for #28099?
> > 
> > 
> > #28099 only has 1 commit. Maybe it is better to reopen this one?
> 
> The reason is that many commits of the old PR have been merged into main([#26210 (comment)](https://github.com/tenstorrent/tt-metal/pull/26210#discussion_r2265553061)). So I only kept the commits that can be used for phi-4 in the new PR.

Ah, I see. Great. Let's work on this first (after that, we could look at phi-1.5 and hopefully it will be also a smaller PR).

**Comment by velonica0** [comment]
> > > > @gwangTT I really appreciate your help with phi-4. Can you help me create a mirror branch for #28099?
> > > 
> > > 
> > > #28099 only has 1 commit. Maybe it is better to reopen this one?
> > 
> > 
> > The reason is that many commits of the old PR have been merged into main([#26210 (comment)](https://github.com/tenstorrent/tt-metal/pull/26210#discussion_r2265553061)). So I only kept the commits that can be used for phi-4 in the new PR.
> 
> Ah, I see. Great. Let's work on this first (after that, we could look at phi-1.5 and hopefully it will be also a smaller PR).

Ok I will try a smaller pPhi-1.5 PR.
Could you please confirm if the next step is to wait for the CI to pass?

Thank you very much for your help again!

**Comment by gwangTT** [comment]
> > > > > @gwangTT I really appreciate your help with phi-4. Can you help me create a mirror branch for #28099?
> > > > 
> > > > 
> > > > #28099 only has 1 commit. Maybe it is better to reopen this one?
> > > 
> > > 
> > > The reason is that many commits of the old PR have been merged into main([#26210 (comment)](https://github.com/tenstorrent/tt-metal/pull/26210#discussion_r2265553061)). So I only kept the commits that can be used for phi-4 in the new PR.
> > 
> > 
> > Ah, I see. Great. Let's work on this first (after that, we could look at phi-1.5 and hopefully it will be also a smaller PR).
> 
> Ok I will try a smaller pPhi-1.5 PR. Could you please confirm if the next step is to wait for the CI to pass?
> 
> Thank you very much for your help again!

Yes, ideally we get all the CI passing (Phi-4 working as bounty required, no regression, etc.).
